### PR TITLE
Add multistage builds to Perseo-FE.

### PR DESCRIPTION
**Pull request by jason-fox**
Separate Node-Slim, Distroless and PM2 Builds.

- Remove entrypoint and  `HEALTHCHECK` dependency on curl.
- Retrieve secrets using the same code as IoT Agent Library (which is Red Hat compatible)
- Improve security by masking sensitive information on start-up.

Tested with the following Docker-Compose

```yaml
version: '3.4'
services:
      # Orion is the context broker
    orion:
        image: fiware/orion:2.5.2
        hostname: orion
        container_name: fiware-orion
        depends_on:
          - perseo-mongo
        networks:
          - default
        expose:
          - "1026"
        ports:
          - "1026:1026" # localhost:1026
        command: -dbhost perseo-mongo -logLevel DEBUG
        healthcheck:
          test: curl --fail -s http://orion:1026/version || exit 1
          interval: 5s

    perseo-mongo:
        image: mongo:4.2
        volumes:
            - ./mongodata:/data/db
        networks:
            - perseo
        deploy:
            replicas: 1
            restart_policy:
                condition: on-failure
    perseo-core:
        image: fiware/perseo-core
        environment:
            - PERSEO_FE_URL=http://perseo-fe:9090
            - MAX_AGE=6000
        networks:
            - perseo
        depends_on:
            - perseo-mongo
        deploy:
            replicas: 1
            restart_policy:
                condition: on-failure
    perseo-fe:
        image: perseo
        networks:
            perseo:
            main:
        ports:
            - 9090:9090
        depends_on:
            - perseo-core
        environment:
            - PERSEO_MONGO_ENDPOINT=perseo-mongo
            - PERSEO_CORE_URL=http://perseo-core:8080
            - PERSEO_LOG_LEVEL=debug
            - PERSEO_ORION_URL=http://orion:1026/
networks:
    perseo:
    main:
```

Result:

```console
docker ps
```

```
174800c7726b   perseo               "docker-entrypoint.s…"   5 minutes ago   Up 4 minutes (healthy)     0.0.0.0:9090->9090/tcp   perseo-fe_perseo-fe_1
55bc55e7e699   fiware/perseo-core   "/code/perseo_core-e…"   5 minutes ago   Up 5 minutes (healthy)     8080/tcp                 perseo-fe_perseo-core_1
552ecfbe44e7   fiware/orion:2.5.2   "/usr/bin/contextBro…"   5 minutes ago   Up 5 minutes (healthy)   0.0.0.0:1026->1026/tcp   fiware-orion
5fd8392c9177   mongo:4.2            "docker-entrypoint.s…"   6 minutes ago   Up 5 minutes               27017/tcp                perseo-fe_perseo-mongo_1
```

### CODE REVIEW

**Review by AlvaroVega — APPROVED** [review]
LGTM

File: docker/Dockerfile
Comment (fgalan) [inline docker/Dockerfile:24]: The change from Node v10 to v12 in the Dockerfile should be noted in CHANGES_NEXT_RELEASE. For instance (inspired in https://github.com/telefonicaid/perseo-fe/releases/tag/1.13.0):

```
- Upgrade NodeJS version from 10.19.0 to 12 in Dockerfile
```

Comment (jason-fox) [reply docker/Dockerfile:24]: Fixed f1dbb24

**Review by fgalan — APPROVED** [review]
LGTM

But let's wait for @IvanHdzC feedback/LGTM before merging.

File: docker/Dockerfile
Comment (IvanHdzC) [inline docker/Dockerfile:87]: @jason-fox is there any reason to download the sources of the app and not use the context sources from where this image is being builded?

I mean, use the `COPY` native function of the Dockerfile and change the build context on the build hook to the root of the repo.

Comment (jason-fox) [reply docker/Dockerfile:87]: Annoyingly `COPY` can't pull from the parent directory. Therefore if the `Dockerfile` isn't in root you have a problem. If the `Dockerfile` is in root then the `README` in the same directory is used on DockerHub - so you still have a problem with broken links in the documentation.

If you think about it, there really is no reason why a `Dockerfile` needs to be in the same repo - many multistage builds start with a _pull external sources from X_ pattern. This happens to use GitHub.

Of course it would be nice to be able to create and build locally with a `COPY` but I've never found a way to do it consistently without amending the base `Dockerfile`.

Comment (IvanHdzC) [reply docker/Dockerfile:87]: I agree with you, there are a lot of cases where pulling external sources is necessary, but I thing this is not the case, mainly because, allways the image is builded with the sources that actually triggered the dockerhub hook. And of course, it's simpler to use, build and understand.

To build the image you can change the build context on the build hook, take [fiware-cygnus Dockerfile](https://github.com/telefonicaid/fiware-cygnus/blob/master/docker/cygnus-ngsi/hooks/build#L19) as an example. As you can see there, on the build hook we change to the root path by doing 

`cd ../../`

In that case we know for sure that the root of the repo is two folders before the Dockerfile path, so we change the current path before building the image. 

It's not fancy but it works fine, and we know for sure how many folders away is the root path of the repo, so we can change the hook in case the Dockerfile location changes.

Comment (jason-fox) [reply docker/Dockerfile:87]: Unfortunately this is not always true. For the TEF Build on DockerHub you are correct, but as a third-party developer looking to amend or contribute to Perseo, I have no access to your build set-up. Indeed I'm probably looking at fiddling with the codebase and maybe raising a PR. 

The point is that my local code base can easily become inconsistent if I don't create from a recognised source of truth (at least TEF mainline by default). If I fork I can pull from my own branched changes and I can also recreate stable tagged releases by setting the build arguments.

```console
docker build -t perseo . --build-arg DOWNLOAD=latest
```

-   You can alter this to obtain the last **stable** release run this `Dockerfile` with the build argument
    `DOWNLOAD=stable`

```console
docker build -t perseo . --build-arg DOWNLOAD=stable
```

-   You can also download a specific release by running this `Dockerfile` with the build argument `DOWNLOAD=<version>`

```console
docker build -t perseo . --build-arg DOWNLOAD=1.7.0
```

## Building from your own fork

To download code from your own fork of the GitHub repository add the `GITHUB_ACCOUNT`, `GITHUB_REPOSITORY` and
`SOURCE_BRANCH` arguments (default `master`) to the `docker build` command.

```console
docker build -t perseo . \
    --build-arg GITHUB_ACCOUNT=<your account> \
    --build-arg GITHUB_REPOSITORY=<your repo> \
    --build-arg SOURCE_BRANCH=<your branch>
    --target=distroless
```


Basically this aligns perseo with the methods used by the IoT Agents

Comment (IvanHdzC) [reply docker/Dockerfile:87]: I see. Regardless of the TEF setup, I can't see the advantage of building the image using args rather than using a different repository/branch/account.

From your point of view. What do you think would be the advantage of using those args than checking out to a different branch or fork?

I mean, using the example of the version 1.7.0. You could run:

```bash
git checkout release/1.7.0
cd .. # To change to the root of the perseo repository. Assuming you are on the docker path of this repo.
docker image build -t perseo -f docker/Dockerfile . # In the case you are on a fork, then you should know where the Dockerfile is located
```

This way seems more native to me.

Comment (jason-fox) [reply docker/Dockerfile:87]: But what if I want to update the baseline image from Ubuntu 18.04 to 20.04? What if I want to create a secure distroless build of 1.7.0, but I don't want to update the source code, just the baseline image?

If I'm deploying new updates I want to minimize the number of moving parts so I can test that change in isolation. I may want  to treat the containerization of the `Dockerfile` and the code version independently.

The idea behind the `build-args` is to give developers flexibility to do what they want (which may differ from your personal workflow).  Of course `COPY .` remains as an option.

Comment (IvanHdzC) [reply docker/Dockerfile:87]: That's part of my point actually. By doing downloading the source code on the dockerfile the `builder` stage must be a debian based image. For example, using alpine as base image would fail. 

```bash
RUN \
	apt-get update && \
	# Ensure that unzip is installed prior to downloading
	apt-get install -y --no-install-recommends unzip ca-certificates curl && \
```

That's the advantage of using the native `COPY` function. It would work on any base distro. And then again, you can allways checkout your workflow to the branch you want to make changes and build from that context.

Anyways. @fgalan & @AlvaroVega what do you think?

Comment (jason-fox) [reply docker/Dockerfile:87]: > `builder` stage must be a debian based image.

Yes, the `builder` stage is however thrown away and not part of your final image. Same would be true of any MUSL based container needing MUSL libs. If you want multistage from alpine you split getting the source code from creating an app - that would be a separate PR of course.

Comment (jason-fox) [reply docker/Dockerfile:87]: @IvanHdzC - this does open a valid question as to which other container variants could/should be supported by the `Dockerfile` - Alpine? RedHat UBI? Additional Architectures beyond ARM? etc. etc.

Comment (IvanHdzC) [reply docker/Dockerfile:87]: This has gone from the original point. Mi original question was to know why the source code is being downloaded on the dockerfile on the builder stage. I'm not questioning the upgrade to multi stage builds on the dockerfile (Which I think it's a great idea). 

I thing we must use native docker/dockerfile instructions in order to prevent failure. From my point of view the dockerfile should allways build from current context. Just as  fiware-cygnus does. If you want to build from a previous version, you can allways change your workflow and build from that context. 

About the aditional architectures I thing with the current builds it's just fine. As I mentioned, my original question was about the build stage. However @fgalan @AlvaroVega what do you thing about this.

Comment (IvanHdzC) [reply docker/Dockerfile:87]: After talking this topic with @fgalan and @AlvaroVega, we've decided to accept this PR as it is. 

However we have oppened an issue to take a deeper look into this topic. You can find it and participate on the issue #498

**Review by IvanHdzC — APPROVED** [review]
LGTM

**Review by fgalan — APPROVED** [review]
LGMT

**Comment by fgalan** [comment]
Travis is failing but it is the usual glitch. Merging anyway.

**Comment by jason-fox** [comment]
> Travis is failing but it is the usual glitch. Merging anyway.

***Cough*** https://github.com/telefonicaid/perseo-fe/pull/470 😃

**Comment by fgalan** [comment]
> 
> 
> > Travis is failing but it is the usual glitch. Merging anyway.
> 
> _**Cough**_ #470 😃

However, as far as as I understand from [this](https://github.com/telefonicaid/perseo-fe/pull/470#issuecomment-736738685), #470 is not ready to merge yet, isn't it?

**Comment by jason-fox** [comment]
@fgalan - #470 is ready for review. The failing test has been fixed

**Comment by fgalan** [comment]
> @fgalan - #470 is ready for review. The failing test has been fixed

Great!

**Pull request merged**
